pykern/xlsx.py

```diff
@@ -0,0 +1,713 @@
+            c._compile_pass2()
+            n += 1
+            # _ROW_MODULUS ensures a gap so columns are separated by more than "+1"
+            # and this will never link the wrong column
+            if p is not None and p.sheet == c.sheet and p.sort_index + 1 == c.sort_index:
+                p = z[-1][1] = c
+                continue
+            z.append([c, None])
+            p = c
+        r = ''
+        for x in z:
+            if r is not None:
+                r += ','
+            r = _xl_id(x[0])
+            if x[1] is not None:
+                r += ':' + _xl_id(x[1])
+        if expect_count is not None and expect_count != n:
+            self._error(
+                'incorrect operands={} expect={} count={}',
+                l[link],
+                expect_count,
+                n,
+            )
+        # _assert_link_pair validates that the cells are the same type
+        return _Operand(
+            cells=l[link],
+            content=r,
+            count=n,
+            fmt=p.get('fmt'),
+            round_digits=p.round_digits,
+            value=p.value if n == 1 else None,
+            is_decimal=p.is_decimal,
+        )
+
+    def _compile_str(self, value):
+        self.round_digits = None
+        self.value = self.content = value
+        self.is_decimal = False
+
+    def _save(self):
+        f = _Fmt(self)
+        self.sheet.width(_XL_COLS[self.col_num], f.width(self))
+        f = self.workbook.xl_fmt(f)
+        if self._is_expr:
+            self.sheet.xl.write_formula(self.xl_id, self.content, value=self.value, cell_format=f)
+        else:
+            self.sheet.xl.write(self.xl_id, self.content, f)
+        if self.is_decimal:
+            self.sheet.text_cells.append(self.xl_id)
+
+
+class _Fmt(PKDict):
+
+    _MAP = PKDict(
+        top=PKDict(top=True),
+        bold=PKDict(bold=True),
+        currency='$#,##0.0',
+        decimal='0.0',
+        percent='0.0%',
+        text='@',
+    )
+    _SPECIAL = re.compile('[$%]')
+    _WIDTH_SLOP = 1
+
+    def width(self, cell):
+        if not cell.is_decimal:
+            return len(cell.value)
+        n = self._WIDTH_SLOP
+        x = cell.value.as_tuple()
+        i = len(x.digits) + x.exponent
+        n += i
+        if ',' in self.num_format:
+            n += i // 3
+        n += cell.round_digits
+        if n and self._SPECIAL.search(self.num_format):
+            n += 1
+        return n
+
+    def __init__(self, cell):
+
+        def _num(fmt):
+            if cell.round_digits is None:
+                return fmt
+            if cell.round_digits == 0:
+                x = ''
+            else:
+                x = '.' + '0' * cell.round_digits
+            fmt = fmt.replace('.0', x)
+            return fmt
+
+        for k in (
+            'font',
+            'border',
+        ):
+            if k in cell.defaults:
+                self.update(self._MAP[cell.defaults[k]])
+        for x in cell.get('fmt', '').split(' '):
```

Comment:
Is this guaranteed to be a single space, or is the usual regexp (```/\s+/```) better?

Reply:
Fix fbaa131